### ISSUE TITLE
bug: subqueries get inlined multiple times, each copy evaluating to different value

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#8619]
### What happened?

Related to https://github.com/ibis-project/ibis/issues/8614

See this example:

```python
import ibis
from ibis import _
from ibis.expr import types as ir

ibis.options.interactive = True


def document_counts(terms: ir.ArrayColumn) -> ir.MapScalar:
    r"""Create a lookup Map from term to number of records containing the term.

    Parameters
    ----------
        terms: One row for each record. Each row is an array of terms in that record.
        Each term could be a word, ngram, or other token from a string.
        Or, it could also represent more generic data, such as a list of tags or
        categories like ["red", "black"]. Each term can be any datatype,
        not just strings.

    Returns
    -------
        A MapScalar with keys as the terms and values as the number of records
        that contain that term.

    Examples
    --------
    >>> import ibis
    >>> from mismo.text import document_counts
    >>> ibis.options.repr.interactive.max_length = 20
    >>> addresses = [
    ...     "12 main st",
    ...     "99 main ave",
    ...     "56 oak st",
    ...     "21 glacier st",
    ...     "12 glacier st",
    ... ]
    >>> t = ibis.memtable({"address": addresses})
    >>> # split on whitespace
    >>> terms = t.address.re_split(r"\s+")
    >>> document_counts(terms)
    {'st': 4, 'ave': 1, '12': 2, '56': 1, '21': 1, 'oak': 1, 'main': 2, 'glacier': 2, '99': 1}
    """  # noqa: E501
    if not isinstance(terms, ir.ArrayValue):
        raise ValueError(f"Unsupported type {type(terms)}")
    terms_table = (
        terms.name("terms")
        .as_table()
        .select(
            record_id=ibis.row_number(),
            terms=_.terms,
        )
    )
    flat = terms_table.select("record_id", term=_.terms.unnest())
    by_term = flat.group_by("term").agg(n_records=_.record_id.nunique())
    # to ensure that .term.collect() and .n_records.collect() are in the same order
    by_term = by_term.view()
    return ibis.map(keys=by_term.term.collect(), values=by_term.n_records.collect())


ibis.options.repr.interactive.max_length = 20
addresses = [
    "12 main st",
    "99 main ave",
    "56 oak st",
    "21 glacier st",
    "12 glacier st",
]
t = ibis.memtable({"address": addresses})
# split on whitespace
terms = t.address.re_split(r"\s+")
dc1 = document_counts(terms)
dc1
# {'main': 2, 'st': 4, 'ave': 1, '21': 1, 'oak': 1, '12': 2, '56': 1, 'glacier': 2, '99': 1}
dc2 = ibis.map(dc1.keys(), dc1.values().map(lambda x: x + 1))
dc2
# oh no the keys and values got shuffled!
# {'glacier': 3, '99': 2, 'st': 2, 'ave': 2, 'main': 3, '12': 2, '56': 3, '21': 5, 'oak': 2}
```

`ibis.to_sql(dc2)` shows

<details>

```sql
SELECT
  MAP(
    MAP_KEYS(
      (
        SELECT
          MAP(ARRAY_AGG("t5"."term"), ARRAY_AGG("t5"."n_records")) AS "Map(ArrayCollect(term), ArrayCollect(n_records))"
        FROM (
          SELECT
            "t3"."term",
            COUNT(DISTINCT "t3"."record_id") AS "n_records"
          FROM (
            SELECT
              "t2"."record_id",
              UNNEST("t2"."terms") AS "term"
            FROM (
              SELECT
                ROW_NUMBER() OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) - CAST(1 AS TINYINT) AS "record_id",
                "t1"."terms"
              FROM (
                SELECT
                  STR_SPLIT_REGEX("t0"."address", '\s+') AS "terms"
                FROM "ibis_pandas_memtable_dwe7rmmmbvcdzlp562lczrxsvu" AS "t0"
              ) AS "t1"
            ) AS "t2"
          ) AS "t3"
          GROUP BY
            1
        ) AS "t5"
      )
    ),
    LIST_APPLY(
      MAP_VALUES(
        (
          SELECT
            MAP(ARRAY_AGG("t5"."term"), ARRAY_AGG("t5"."n_records")) AS "Map(ArrayCollect(term), ArrayCollect(n_records))"
          FROM (
            SELECT
              "t3"."term",
              COUNT(DISTINCT "t3"."record_id") AS "n_records"
            FROM (
              SELECT
                "t2"."record_id",
                UNNEST("t2"."terms") AS "term"
              FROM (
                SELECT
                  ROW_NUMBER() OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) - CAST(1 AS TINYINT) AS "record_id",
                  "t1"."terms"
                FROM (
                  SELECT
                    STR_SPLIT_REGEX("t0"."address", '\s+') AS "terms"
                  FROM "ibis_pandas_memtable_dwe7rmmmbvcdzlp562lczrxsvu" AS "t0"
                ) AS "t1"
              ) AS "t2"
            ) AS "t3"
            GROUP BY
              1
          ) AS "t5"
        )
      ),
      __ibis_param_x__ -> __ibis_param_x__ + CAST(1 AS TINYINT)
    )
  ) AS "Map(MapKeys(ScalarSubquery()), ArrayMap(MapValues(ScalarSubquery()), Add(x, 1)))"
```

</details>

Note how the expression for dc1 is copied in two places. Thus when evaluated, the two places are not guaranteed to have the same value.

Note how I am sometimes able to get around this, for instance the `by_term = by_term.view()` inside the `document_counts()` function. But I haven't managed to come up with a way to "singleton-ize" dc1.

The SQL that I want, and that I think most users would expect, would be for dc1 to get turned into a CTE, so that all subsequent references refer to the same value.

I think this is also a problem for all sorts of subqueries, not just MapScalars. Tables, that rely on sort order, etc.

### What version of ibis are you using?

main

### What backend(s) are you using, if any?

duckdb

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Thanks!

This is definitely something we want to address, and it's related to some of the other performance issues you observed.

**Comment by kszucs** [comment]
This is most certainly because we are not traversing into value expressions while extracting CTEs, see the passed types to the traversal `filter`:

https://github.com/ibis-project/ibis/blob/9b006572fafc868b0cd45582a4e872466b93b022/ibis/backends/sql/rewrites.py#L184-L195

Perhaps not limiting that would resolve it without side effects.

**Comment by cpcloud** [comment]
Closed by #8633

**Comment by NickCrews** [comment]
@cpcloud this bug is actually still present, if you try my original repro sample above
